### ISSUE TITLE
CI to build and deploy the lastest site

### ISSUE THREAD

**Issue by ketzacoatl** [haskell-infra/www.haskell.org#66]
In #57, it is pointed out that the date will auto-correct the next time the site is built/deployed. While it seems there is an existing `.gitlab-ci.yml`, it doesn't seem to include anything for publishing the updates that have been merged. Considering this is a static site built with hakyll, it ought to be relatively easy to add this, but maybe there are additional TODOs to integrate with how the site is hosted.

### requirements

* Updates that have been validated and merged should be built and deployed by CI without requiring an admin taking action (approval or triggers are reasonable, but this update process should not require an admin uploading files or similar).
* CI should deploy whatever is on `master`, so additional updates to `master` (including reverting a change) update the live deployment.

**Comment by TikhonJelvis** [comment]
Yeah, this is a great idea. Currently deploying is manual and a bit fiddly—fixing that is definitely on our todo list for this year.

In general, I think reducing friction for making improvements to the site will help in keeping it up to date and improving it incrementally.

**Comment by ketzacoatl** [comment]
@TikhonJelvis I have a lot of experience with CI/CD, maybe I can help get this auto-deploy in place? This is a static site, so It should be really easy todo.

Can you confirm what the deployment should be made to?
Is there a server we have to SSH into, or a cloud provider to hit an API on?
How much should the current setup change as part of this (or not), and in what ways?

**Comment by TikhonJelvis** [comment]
Currently we SSH into the server that hosts the site and rebuild it in place. That's fine for now, but with CI a better solution would be to build the site as part of the CI script, run some error-checking (like `linkcheck` that's already run by `nix-build`) and then upload the results to the hosting server.

I'm absolutely happy to help on the server and Nix end of things, but I'm not familiar with setting up GitHub actions or managing secrets like the SSH key for deployment.

**Comment by ketzacoatl** [comment]
@TikhonJelvis do we _need_ to use a server? Could we possibly use GitHub static pages (and point www.haskell.org DNS at that). Also, is there only _one_ server? Could we consider some HA here?

**Comment by bgamari** [comment]
Note that ghc.haskell.org is currently served via gitlab pages; Feel free to borrow anything you find useful from https://gitlab.haskell.org/ghc/homepage.

**Comment by gbaz** [comment]
I would generally tend to oppose serving things via github static pages given that we have an infrastructure already running. This just makes a free software project more dependent on the whims of a large corporation. Given that we necessarily have to run a fair amount of our own infrastructure, there's no real pain point to having this site be served as well.

**Comment by ketzacoatl** [comment]
I 100% agree on the dilema of relying on GitHub for that hosting. However, I also believe that a change like this will *greatly* improve not only our experience contributing to the haskell.org website, but also the reliability and hosting of that content. 

It's worth noting, we're not limited to GitHub, we could use a variety of service providers for this content, https://www.digitalocean.com/products/app-platform/ is a worthy contender. *It is just a static website.* So I don't see a reason to get tripped up on the GitHub part of this - the point is to improve and evolve the infrastructure and hosting for the haskell community. @gbaz, what personal investment to you have in all this stuff? You seem to be uninterested in reasonable improvements to infrastructure and processes you are involved in. What do you need to protect?

> there's no real pain point to having this site be served as well.

There are numerous problems that these small improvements could start to address:
* @TikhonJelvis has been blocked on the nuances of accessing the existing server, probably others have as well.
* PRs get merged but deploys have to wait until a person with access can hit the keyboard.
* People who are competent and capable cannot help support the infrastructure until they have been approved by you thru some opaque process, whereas we could instead have a platform for hosting (managed you) but that does not block or interfere progress like it does today.
* The hosting setup is opaque, and there's no way for a contributor to help with improving the CI/CD for the project
* There is only one server.
* I'll say it again, because it is so ridiculous in 2021 to be saying this: there is only one server.
* Reviewing haskell history, admins have made mistakes and both irresponsibly and inadvertently deleted production servers running haskell infrastructure.
* AFAICT, there are no "requirements" or "expectations" defined for hosting infrastructure. There is little on that in the hosting documentation, and we don't have any code defining these deployments, so a contributor like me cannot validate that backups are happening, for example, or that there is a reasonable hosting strategy.
* We could easily have *code* which is responsible for the deployment, and which makes it trivial to re-deploy, update, etc.

In 2021, there is no excuse for blocking or deferring progress towards easily reproducible deployments based on code and using CI/CD/etc tooling.

As the platform/infrastructure team, you are supposed to enable and support developer and contributors, not block or get in their way on easily resolved issues.

@gbaz, do you see yourself as the gatekeeper on reasonable discussion and collaboration towards improving haskell? AFAICT, you are a significant cause of burnout in this community, and it has got to stop.

**Comment by gbaz** [comment]
Please take that chip off your shoulder. It can't be very pleasant to carry it around.

**Comment by ketzacoatl** [comment]
@TikhonJelvis, @rebeccaskinner, what level of interest does the committee have in this idea of CI and an improved build/publish workflow? FWIW, I believe the current setup could be improved significantly with a small investment in defining how it'll work and making it happen. Whether it's on GitHub, or GitLab, I would be happy to contribute the code to implement. But I'm not going to invest that effort without coming to some sort of consensus or agreement on what that thing is and the desire for it.

**Comment by TikhonJelvis** [comment]
Hey, yeah, CI is definitely a good idea, and thanks for offering to help with it. We need to figure out how to manage the server side (SSH keys, a deploy script) before CI makes sense. That's on my to-do list for the next month or two, I just haven't had the time/energy to get to it over the last few weeks.

**Comment by ketzacoatl** [comment]
It doesn't need to be a lot of work. For one thing, as a static site, it'd be a _lot_ easier to push that rendered site to a service which does the hosting. We could use GitHub, GitLab, or one of the few paid CDN/static site hosting services. That change alone would significantly reduce the work compared to the current, single-pet-server hosting. We could also more easily document the details of the setup. If there is too much resistance to improving the single server setup, we can figure how to keep it simple.

>  We need to figure out how to manage the server side (SSH keys, a deploy script) before CI makes sense.

Can you elaborate some more on this? What is there to figure out? The deploy requires SSH access and to scp/rsync the files, no? The SSH key can be stored securely in CI, the script would live in this repo. I can understand if there is more complication I'm not seeing, and I'd be happy to help navigate that. Again, I believe the small bit of time invested here will payout many times over as updates are made to the site.

**Comment by tomjaguarpaw** [comment]
There is a PR adding CI to build and check links at https://github.com/haskell-infra/www.haskell.org/pull/76. I look forward to getting the opportunity to add other functionality on top.

**Comment by tomjaguarpaw** [comment]
If we envisage haskell.org remaining a static site then I am strongly in favour of serving it on a hosting provider that provides exactly that service and requires no extra configuration or maintenance. However, this desire is complicated by the fact that not everything on haskell.org comes from this repository (in fact it's not even close).

I started a [section of the README](https://github.com/haskell-infra/www.haskell.org#subsites) to document the sources of other subsites. You can see the link to the cabal subsite repo there. There also used to be a link to the haskell-platform subsite repo (it was removed, [perhaps due to a misunderstanding](https://github.com/haskell-infra/www.haskell.org/issues/77)). There are several other subsites not listed, including alex, arrows, and hoogle. When I get a chance I will list them in a public document and try to determine who is responsible for each of them so that we can discuss and build consensus about what to do.

**Comment by ketzacoatl** [comment]
>  However, this desire is complicated by the fact that not everything on haskell.org comes from this repository (in fact it's not even close).

That's easy to overlook, thanks for taking on the effort to document some of those details.

Based on experience with these things, I would start off considering the following:
* Using the CDN or similar routing features in static hosting, each project is build/published/hosted separately - CI for each project publishes their artifacts separately, and the CDN handles the path routing.
* The CI for this project could build the resources from those dependent projects, and include them in the final site.
* CI for those dependent projects could publish artifacts, which this project's CI then consumes.
* Artifacts for those projects could be built manually, and published in a way this project's CI could then consume and use.

**Comment by tomjaguarpaw** [comment]
> Based on experience with these things, I would start off considering the following

Much appreciated. I'm supportive of that course of action.

**Comment by ketzacoatl** [comment]
Closing as a duplicate of #83.
